librad/src/git/trailer.rs

```diff
@@ -0,0 +1,256 @@
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+#[derive(Debug, Clone, Eq, PartialEq, thiserror::Error)]
+pub enum Error<'a> {
+    #[error("The trailers paragraph is missing in the given message")]
+    MissingParagraph,
+
+    #[error(
+        "One or more trailers are not in the parseable format <token><separator><value>: '{0}'"
+    )]
+    Unparsable(&'a str),
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct Trailer<'a> {
+    token: Token<'a>,
+    values: Vec<&'a str>,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct Token<'a>(&'a str);
+
+/// Parse the trailers of the given message. It looks up the last paragraph
+/// of the message and attempts to parse each of its lines as a [Trailer].
+/// Fails if no trailers paragraph is found or if at least one trailer
+/// fails to be parsed.
+pub fn parse<'a>(message: &'a str, separators: &'a str) -> Result<Vec<Trailer<'a>>, Error<'a>> {
+    let trailers_paragraph =
+        match parser::paragraphs(message.trim_end()).map(|(_, ps)| ps.last().map(Clone::clone)) {
+            Ok(None) | Err(_) => return Err(Error::MissingParagraph),
+            Ok(Some(p)) => p,
+        };
+
+    match parser::trailers(trailers_paragraph, separators) {
+        Ok((rest, trailers)) if rest.is_empty() => Ok(trailers),
+        Ok((unparseable, _)) => Err(Error::Unparsable(unparseable)),
+        Err(_) => Err(Error::Unparsable(trailers_paragraph)),
+    }
+}
+
+pub mod parser {
+    use super::{Token, Trailer};
+    use nom::{
+        branch::alt,
+        bytes::complete::{tag, take_until, take_while1},
+        character::complete::{line_ending, not_line_ending, one_of, space0, space1},
+        combinator::rest,
+        multi::{many0, separated_list},
+        sequence::{delimited, preceded, separated_pair, terminated},
+        IResult,
+    };
+
+    const EMPTY_LINE: &str = "\n\n";
+
+    pub fn paragraphs(s: &str) -> IResult<&str, Vec<&str>> {
+        separated_list(tag(EMPTY_LINE), paragraph)(s)
+    }
+
+    pub fn paragraph(s: &str) -> IResult<&str, &str> {
+        alt((take_until(EMPTY_LINE), rest))(s)
+    }
+
+    /// Parse all the possible trailers.
+    /// It stops when it can no longer parse valid trailers.
+    pub fn trailers<'a>(s: &'a str, separators: &'a str) -> IResult<&'a str, Vec<Trailer<'a>>> {
+        many0(|s| trailer(s, separators))(s)
+    }
+
+    /// Parse a trailer, which can have an inlined or multilined value.
+    pub fn trailer<'a>(s: &'a str, separators: &'a str) -> IResult<&'a str, Trailer<'a>> {
+        let parser = separated_pair(token, |s| separator(separators, s), values);
+        let (rest, (token, values)) = parser(s)?;
+        Ok((rest, Trailer { token, values }))
+    }
+
+    /// Parse a trailer token.
+    fn token(s: &str) -> IResult<&str, Token> {
```

Comment:
Curious as to why these are private and not `pub`?

Reply:
No particular reason didn't _need_ to make them `pub`.